plugins/mediatek-scaler/fu-mediatek-scaler-device.c

```diff
@@ -663,89 +702,93 @@ fu_mediatek_scaler_device_chunk_data_is_blank(FuChunk *chk)
+	if (fu_mediatek_scaler_device_chunk_data_is_blank(helper->chk)) {
+		/* fast forward if chunk is empty */
+		if (!fu_mediatek_scaler_device_set_data_fast_forward(self, helper->sent_sz, error))
 			return FALSE;
-		}
+	} else {
+		/* set data per fragment size */
+		if (!fu_mediatek_scaler_device_set_data(self, helper->chk, error))
+			return FALSE;
+	}
 
-		g_debug("retry write_firmware: step: %d, max: %d", retry, DDC_RW_MAX_RETRY_CNT);
+	/* verify the sent data chunk */
+	if (!fu_mediatek_scaler_device_check_sent_info(self, helper->chk, helper->sent_sz, error)) {
+		/* restore the data size counter */
+		if (!fu_mediatek_scaler_device_set_data_fast_forward(
+			self,
+			helper->sent_sz - fu_chunk_get_data_sz(helper->chk),
+			error))
+			return FALSE;
 	}
-	g_set_error_literal(error, FWUPD_ERROR, FWUPD_ERROR_INTERNAL, "maximum tries exceeded");
-	return FALSE;
+
+	/* ff to reset the checksum */
+	if (!fu_mediatek_scaler_device_set_data_fast_forward(self, helper->sent_sz, error))
+		return FALSE;
+
+	return TRUE;
 }
 
 static gboolean
-fu_mediatek_scaler_device_attach(FuDevice *device, FuProgress *progress, GError **error)
+fu_mediatek_scaler_device_write_firmware_impl(FuMediatekScalerDevice *self,
+					      GInputStream *stream,
+					      FuProgress *progress,
+					      GError **error)
 {
-	FuMediatekScalerDevice *self = FU_MEDIATEK_SCALER_DEVICE(device);
-	guint max_tries = 30;
+	guint32 sent_sz = 0x0;
+	g_autoptr(FuChunkArray) chunks = NULL;
 
-	/* reboot the device */
-	if (!(fu_mediatek_scaler_device_set_isp_reboot(self, error)))
+	chunks = fu_chunk_array_new_from_stream(stream, 0x00, DDC_DATA_PAGE_SIZE, error);
+	if (chunks == NULL)
 		return FALSE;
 
-	/* wait for the device back */
-	if (!fu_device_retry_full(device,
-				  fu_mediatek_scaler_device_display_is_connected_cb,
-				  max_tries,
-				  FU_MEDIATEK_SCALER_DEVICE_POLL_INTERVAL,
-				  NULL,
-				  error)) {
-		g_prefix_error(error,
-			       "display controller did not reconnect after %u retries: ",
-			       max_tries);
-		return FALSE;
+	for (guint i = 0; i < fu_chunk_array_length(chunks); i++) {
+		FuMediatekScalerWriteChunkHelper helper_wchunk = {0x0};
+		g_autoptr(FuChunk) chk = NULL;
+
+		/* prepare chunk */
+		chk = fu_chunk_array_index(chunks, i, error);
+		if (chk == NULL)
+			return FALSE;
+
+		/* data size already sent to chip */
+		sent_sz += fu_chunk_get_data_sz(chk);
+
+		/* retry writing data chunk */
+		helper_wchunk.chk = chk;
+		helper_wchunk.sent_sz = sent_sz;
+		if (!fu_device_retry_full(FU_DEVICE(self),
+					  fu_mediatek_scaler_device_write_chunk,
+					  DDC_RW_MAX_RETRY_CNT,
+					  FU_MEDIATEK_SCALER_DDC_MSG_DELAY_MS,
+					  &helper_wchunk,
+					  error)) {
+			g_prefix_error(error, "writing chunk exceeded the maximum retries");
+			return FALSE;
+		}
+
+		/* write chunk successfully, update the progress */
+		fu_progress_set_percentage_full(fu_progress_get_child(progress),
+						(gsize)i + 1,
+						(gsize)fu_chunk_array_length(chunks));
```

Comment:
I think you should do `fu_progress_set_steps(progress, fu_chunk_array_length(chunks))` above and then `fu_progress_done()` here. That way the daemon knows the number of steps it's expecting.

Reply:
fixed.